pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto

```diff
@@ -178,14 +178,39 @@ message Mount {
     MountPropagation propagation = 5;
 }
 
+// A NamespaceMode describes the intended namespace configuration for each
+// of the namespaces (Network, PID, IPC) in NamespaceOption. Runtimes should
+// map these modes as appropriate for the technology underlying the runtime.
+enum NamespaceMode {
+    // A POD namespace is common to all containers in a pod.
+    // For example, a container with a PID namespace of POD expects to view
+    // all of the processes in all of the containers in the pod.
+    POD       = 0;
+    // A CONTAINER namespace is restricted to a single container.
+    // For example, a container with a PID namespace of CONTAINER expects to
+    // view only the processes in that container.
+    CONTAINER = 1;
+    // A NODE namespace is the namespace of the Kubernetes node.
+    // For example, a container with a PID namespace of NODE expects to view
+    // all of the processes on the host running the kubelet.
+    NODE      = 2;
+}
+
 // NamespaceOption provides options for Linux namespaces.
 message NamespaceOption {
-    // If set, use the host's network namespace.
-    bool host_network = 1;
-    // If set, use the host's PID namespace.
-    bool host_pid = 2;
-    // If set, use the host's IPC namespace.
-    bool host_ipc = 3;
+    // Network namespace for this container/sandbox.
+    // Note: There is currently no way to set CONTAINER scoped network in the Kubernetes API.
+    // Namespaces currently set by the kubelet: POD, NODE
+    NamespaceMode network = 1;
```

Comment:
@yujuhong I don't know enough about proto wire format to know if changing these types will cause an error on deserialize (which is what I want) or garbage bools. Any ideas?

Reply:
I think it might not break, but it'd be good to verify the behavior either way.

Reply:
It breaks since the filed name/type are changed with the same id. If you add those with new ids (e.g. 4,5,6), then it won't break.

Reply:
Researched a bit here and tested using the remote dockershim. It turns out that [proto3 fields are compatible across types](https://developers.google.com/protocol-buffers/docs/proto3#updating), including bools & enums, so we can't just change the type. The behavior appears to be that the enum will be cast to a bool resulting in lots of containers getting host namespaces.

Instead we can renumber the fields. The old bools will get default false values. The effect is that host network/pid/IPC will stop working until the runtimes update their proto.

Reply:
I don't now if there is a cleaner way to resolve this short of kubelet performing a version check (which will come later).
Defaulting to POD namespace mode does sound safer.

/cc @Random-Liu @mrunalp

Reply:
@yujuhong, we could also do something like this:

```
message NamespaceOption {                                                                                                                                                        
    // Prior to using NamespaceMode, NamespaceOption used bools to specify namespace.                                                                                            
    // This unused field is type-incomptabile with the previous bool in order to force                                                                                           
    // an error when a runtime implements the outdated interface.                                                                                                                
    float sentinel = 1;                                                                                                                                                          
    reserved 2 to 3;   
...
```

Then the kubelet would always set `sentinel` to a non-zero which will cause a grpc error and no containers will start:

```
I0204 22:14:02.194405   11572 kuberuntime_manager.go:579] SyncPod received new pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)", will create a sandbox for it       
I0204 22:14:02.194427   11572 kuberuntime_manager.go:588] Stopping PodSandbox for "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)", will start new one                  
I0204 22:14:02.194449   11572 kuberuntime_manager.go:640] Creating sandbox for pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)"                                     
E0204 22:14:02.195188   11572 remote_runtime.go:92] RunPodSandbox from runtime service failed: rpc error: code = Internal desc = grpc: error unmarshalling request: proto: wrong 
wireType = 5 for field HostNetwork                                                                                                                                               
E0204 22:14:02.195262   11572 kuberuntime_sandbox.go:54] CreatePodSandbox for pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)" failed: rpc error: code = Internal de
sc = grpc: error unmarshalling request: proto: wrong wireType = 5 for field HostNetwork                                                                                          
E0204 22:14:02.195281   11572 kuberuntime_manager.go:646] createPodSandbox for pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)" failed: rpc error: code = Internal d
esc = grpc: error unmarshalling request: proto: wrong wireType = 5 for field HostNetwork                                                                                         
E0204 22:14:02.195340   11572 pod_workers.go:186] Error syncing pod 49c4f606-09f0-11e8-b5dc-42010af00002 ("debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)"), skipping: f
ailed to "CreatePodSandbox" for "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)" with CreatePodSandboxError: "CreatePodSandbox for pod \"debugtest_default(49c4f606-09f0
-11e8-b5dc-42010af00002)\" failed: rpc error: code = Internal desc = grpc: error unmarshalling request: proto: wrong wireType = 5 for field HostNetwork"                         
I0204 22:14:02.195449   11572 server.go:423] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"debugtest", UID:"49c4f606-09f0-11e8-b5dc-42010af00002", APIVersion:"
v1", ResourceVersion:"271", FieldPath:""}): type: 'Warning' reason: 'FailedCreatePodSandBox' Failed create pod sandbox: rpc error: code = Internal desc = grpc: error unmarshalli
ng request: proto: wrong wireType = 5 for field HostNetwork 
```